5 Object Oriented Programming/OOP_readme.md

```diff
@@ -0,0 +1,122 @@
+Object-oriented programming, a set of techniques that use objects (and related concepts) as the central idea of program organization.
+
+**Encapsulation**
+
+The core idea in object-oriented programming is to divide programs into smaller pieces and make each piece responsible for managing its own state. This way, some knowledge about the way a piece of the program works can be kept local to that piece. Someone working on the rest of the program does not have to remember or even be aware of that knowledge. Whenever these local details change, only the code directly around it needs to be updated. Different pieces of such a program interact with each other through interfaces, limited sets of functions or bindings that provide useful functionality at a more abstract level, hiding their precise implementation. Such program pieces are modelled using objects. Their interface consists of a specific set of methods and properties. Properties that are part of the interface are called public. The others, which outside code should not and cannot touch, are called private. Even though the language doesn&#39;t have this distinction built in, JavaScript programmers are successfully using this idea. Typically, the available interface is described in documentation or comments. It is also common to put an underscore (\_) character at the start of property names to indicate that those properties are private. The idea of separating interface from implementation is called encapsulation.
+
+**Methods**
+
+Methods are nothing more than properties that hold function values. This is a simple method:<br>
+let rabbit = {};<br>
+	&nbsp;rabbit.speak = function(line) {<br>
+	&nbsp;console.log(`The person says '${line}'`);<br>
+};
+
+rabbit.speak("I'm good.");    //The rabbit says I'm good.
+
+Usually a method needs to do something with the object it was called on when a function is called as a method— looked up as a property and immediately called, as in object.method()— the binding called this in its body automatically points at the object that it was called on.
+
+function speak(line) {<br>
+&npsp;console.log(`The ${this.type} rabbit says '${line}'`);<br>
+}<br>
+let whiteRabbit = {<br>
+    &nbsp;type: "white",<br>
+    &nbsp;speak<br>
+};<br>
+let hungryRabbit = {<br>
+    &nbsp;type: "hungry",<br>
+    &nbsp;speak<br>
+};<br>
+whiteRabbit.speak("Oh my ears and whiskers, " + "how late it's getting!"); <br>     // The white rabbit says 'Oh my ears and whiskers, how late it's getting!' <br>
+hungryRabbit.speak("I could use a carrot right now.");     //The hungry rabbit says 'I could use a carrot right now.'   <br>
+
+You can think of this as an extra parameter that is passed in a different way. If you want to pass it explicitly, you can use a function’s call method, which takes the ‘this’ value as its first argument and treats further arguments as normal parameters.<br><br>
+speak.call(hungryRabbit, "Tasty!");<br>

```

Comment:
# Object-Oriented Programming
Object-oriented programming, a set of techniques that use objects (and related concepts) as the central idea of program organization.
## Key Concepts:

**Encapsulation**

The core idea in object-oriented programming is to divide programs into smaller pieces and make each piece responsible for managing its own state. This way, some knowledge about the way a piece of the program works can be kept local to that piece. Someone working on the rest of the program does not have to remember or even be aware of that knowledge. Whenever these local details change, only the code directly around it needs to be updated. Different pieces of such a program interact with each other through interfaces, limited sets of functions or bindings that provide useful functionality at a more abstract level, hiding their precise implementation. Such program pieces are modelled using objects. Their interface consists of a specific set of methods and properties. Properties that are part of the interface are called public. The others, which outside code should not and cannot touch, are called private. Even though the language doesn&#39;t have this distinction built in, JavaScript programmers are successfully using this idea. Typically, the available interface is described in documentation or comments. It is also common to put an underscore (\_) character at the start of property names to indicate that those properties are private. The idea of separating interface from implementation is called encapsulation.

**Methods**

Methods are nothing more than properties that hold function values. This is a simple method:
````js
let rabbit = {};<br>
	rabbit.speak = function(line) {
	console.log(`The person says '${line}'`);<br>
};

rabbit.speak("I'm good.");    //The rabbit says I'm good.
````
Usually a method needs to do something with the object it was called on when a function is called as a method— looked up as a property and immediately called, as in object.method()— the binding called this in its body automatically points at the object that it was called on.
```js
function speak(line) {
console.log(`The ${this.type} rabbit says '${line}'`);
}
let whiteRabbit = {
    type: "white",
    speak
};
let hungryRabbit = {
    type: "hungry",
    speak
};

whiteRabbit.speak("Oh my ears and whiskers, " + "how late it's getting!"); <br>     // The white rabbit says 'Oh my ears and whiskers, how late it's getting!' <br>
hungryRabbit.speak("I could use a carrot right now.");     //The hungry rabbit says 'I could use a carrot right now.'   <br>

You can think of this as an extra parameter that is passed in a different way. If you want to pass it explicitly, you can use a function’s call method, which takes the ‘this’ value as its first argument and treats further arguments as normal parameters.<br><br>
speak.call(hungryRabbit, "Tasty!");<br>
```

Reply:
**That is how you should be formatting your markdown.**